ls1intum/Artemis
# `Assessment`:  Fix issues during modeling assessments in tutor training

### Checklist
#### General


#### Client

### Motivation and Context
This is a follow up PR that introduces a client test for the bugfix [#4940](https://github.com/ls1intum/Artemis/pull/4940) and fixes the duplicate request causing `DataIntegrityViolationException` (displayed to the user as `Internal server error`) and the error causing the example submission for modeling exercises to be displayed as empty during tutor assessment training. The relevant Sentry entry for the `DataIntegrityViolationException` is [here](https://sentry.ase.in.tum.de/organizations/artemis/issues/21770/events/1be9ec57e7f244e7b552e5aec266ca29/?project=2).

Another error this PR addresses is that when the UML elements/relations for a modeling exercise example submission is deleted, the referencing feedback is not deleted so there were dangling points that is present in the example submissions list page but no visible feedback is associated with them. This was leading to confusion and make it seem like the model is not loaded properly.

### Description
In order to prevent the same request to create the tutor participation after the tutor reads the instructions, the button is disabled while the response is loading.

Other issue causing the modeling exercises to be displayed as empty during tutor assessment training is due to a race condition. It occurs when the `exercise` is set before the `umlModel` of the example submission in `ExampleModelingSubmissionComponent`. Two independent observables are synchronized to prevent the race condition.

For the dangling points and feedbacks issue, we now delete the reference feedbacks which does not reference an existing UML element/relation.

### Steps for Testing

Prerequisites:
- 1 Tutor
- 1 Instructor
- 1 Modeling Exercise with Example Submission

1. Log in to Artemis with the Tutor account
2. Navigate to Course Administration
3. Go to the course of the prerequisite Modeling Exercise > Assessment Dashboard > Modeling Exercise's Exercise Dashboard
4. Click `Start assessing example submissions` multiple times and check that the button is disabled after the first one and no Internal server error notification is shown (You may need to enable throttling at browser DevTools because it doesn't usually take too long to finish loading).
6. Check that the example solution, assessment instructions and `Submit assessment` button is visible and the model is not empty.
8. Make an assessment and then submit to see the result.

9. Log in to Artemis with the Instructor account
10. Navigate to Course Administration
11. Go to the course of the prerequisite Modeling Exercise > Example Submissions
12. Create new or edit an existing example submission
13. Add some UML elements, Save Example Submission and then click Show Assessments.
14. Add some feedback on the UML elements/associations and then Save Example Assessment
15. Check Modeling Exercise > Example Submissions list page displays total points correctly.
16. Click Show Submissions button and delete a UML element/association with feedback, then Save Example Submission.
17. Check Modeling Exercise > Example Submissions list page displays total points correctly (The point for feedback referencing the deleted UML element/association should be removed).

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
| example-text-submission.component.ts  | 78.7% | 95.87% |
| exercise-assessment-dashboard.component.ts  | 78.6% | 95.12% |
| example-modeling-submission.component.ts | 62.69% | 92.2% |

### Screenshots



Before the fix (not always, indeterministic):
![image](https://user-images.githubusercontent.com/18427209/164758590-a417fff7-74f4-49ab-ba39-c9ad57d793b4.png)


After the fix:
![image](https://user-images.githubusercontent.com/18427209/164758630-317b890c-829d-4e62-8c24-8d43d3ac71ff.png)
